Be able to skip CSV Header in CSV export

Shoprunners opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | To give the CSVResponse functionality a bit more flexibility, this PR add the possibility to skip header rows.
| Type?         | improvement
| Category?     |  CO
| BC breaks?    | yes
| Deprecations? | no
| Fixed ticket? | Fixes #28073
| How to test?  | create a module which uses CSVResponse Class to create CSV File and call setExportHeader(false). This should skip adding the header row. Module [csvexporttest.zip](https://github.com/PrestaShop/PrestaShop/files/8408816/csvexporttest.zip)

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/28072)
<!-- Reviewable:end -->

Shoprunners committed:
Be able to skip CSV Header in CSV export

Shoprunners commented:
added an issue

PierreRambaud reviewed (dismissed):
Unfortunately, the CI is broken

Shoprunners commented:
Yes, but I don't see why. Can you help?

PierreRambaud commented:
> Yes, but I don't see why. Can you help?

Everything is defined here: 
![image](https://user-images.githubusercontent.com/1462701/160785015-dee1610a-92e1-49e6-b196-4d70f232cc9a.png)

Shoprunners commented:
> > Yes, but I don't see why. Can you help?
> 
> Everything is defined here: ![image](https://user-images.githubusercontent.com/1462701/160785015-dee1610a-92e1-49e6-b196-4d70f232cc9a.png)

Yes, I know, but there is no error marked. I cannot interpret the error message, what does it say?

Progi1984 commented on src/PrestaShopBundle/Component/CsvResponse.php:
```diff
@@ -166,6 +171,18 @@ public function setFileName($fileName)
+     *
+     * @return $this
+     */
+    public function setExportHeader($exportHeader)
```
```suggestion
    public function setExportHeader(bool $exportHeader): self
```

Progi1984 commented on src/PrestaShopBundle/Component/CsvResponse.php:
```diff
@@ -67,6 +67,11 @@ class CsvResponse extends StreamedResponse
     private $limit = 1000;
 
+    /**
+     * @var boolean exportHeader
```
```suggestion
     * @var bool exportHeader
```

Progi1984 commented on src/PrestaShopBundle/Component/CsvResponse.php:
```diff
@@ -166,6 +171,18 @@ public function setFileName($fileName)
     }
 
+    /**
+     * @param boolean $exportHeader
```
```suggestion
     * @param bool $exportHeader
```

Progi1984 commented on src/PrestaShopBundle/Component/CsvResponse.php:
```diff
@@ -166,6 +171,18 @@ public function setFileName($fileName)
+     */
+    public function setExportHeader($exportHeader)
+    {
+        $this->exportHeader = (boolean) $exportHeader;
```
```suggestion
        $this->exportHeader = $exportHeader;
```

Shoprunners committed:
Update src/PrestaShopBundle/Component/CsvResponse.php

Co-authored-by: Progi1984 <progi1984@gmail.com>

Shoprunners committed:
Update src/PrestaShopBundle/Component/CsvResponse.php

Co-authored-by: Progi1984 <progi1984@gmail.com>

Progi1984 reviewed (dismissed):
@Shoprunners Thank you for your contribution. Could you fix my feedbacks and remove your merge commit ?

Shoprunners committed:
Update src/PrestaShopBundle/Component/CsvResponse.php

Co-authored-by: Progi1984 <progi1984@gmail.com>

Shoprunners committed:
Update src/PrestaShopBundle/Component/CsvResponse.php

Co-authored-by: Progi1984 <progi1984@gmail.com>

atomiix commented:
@Shoprunners Could you provide us a test module so it will be easier for the QA team to validate your PR? :)

Shoprunners commented:
Is this sufficient?
[csvexporttest.zip](https://github.com/PrestaShop/PrestaShop/files/8408816/csvexporttest.zip)

matthieu-rolland commented:
thank you @Shoprunners for your your pull request and the test module !

I see that the product team doesn't have validated your addition yet, once they do we can send your PR in QA 👍

atomiix commented:
@Shoprunners thanks for the test module, it should do the job 👍

matks commented:
@PrestaShop/product-team Could you tell us if https://github.com/PrestaShop/PrestaShop/issues/28073 is valid so we can go forward for this PR? 😊

fabienvallon commented:
LGTM

Progi1984 reviewed (changes requested):
It miss a getter for `exportHeader` property.

Shoprunners commented:
> It miss a getter for `exportHeader` property.

Hi, I see the point, but no private var at all has a getter method in this class. Do you really want me to add one? Actually a CSVResponse object little to no chance to be reused ?

Progi1984 commented:
@Shoprunners Yes, it's better to have a getter to know the status of this boolean property.

Shoprunners committed:
Update CsvResponse.php

1. Adds a new getter method to retrieve the state of including the header row or not
2. Rename the variables to beeing more intuitive.

Shoprunners commented:
Okay, I added the getter and renamed the variable in general.

Shoprunners committed:
Update CsvResponse.php

bool instead of boolean in comment section

Shoprunners committed:
Update CsvResponse.php

CS Fixer

Progi1984 commented on src/PrestaShopBundle/Component/CsvResponse.php:
```diff
@@ -86,6 +91,16 @@ public function __construct($callback = null, $status = 200, $headers = [])
+     *
+     * @return bool
+     */
+    public function getIncludeHeaderRow()
```
```suggestion
    public function isHeaderRowIncluded(): bool
```

Progi1984 reviewed (commented):
A small change in naming

Shoprunners committed:
Update src/PrestaShopBundle/Component/CsvResponse.php

Co-authored-by: Progi1984 <progi1984@gmail.com>

atomiix commented:
@Shoprunners you have a merge commit in your git history. Can you rebase your code and get rid of it? :)

Shoprunners commented:
> @Shoprunners you have a merge commit in your git history. Can you rebase your code and get rid of it? :)

Not sure... is it possible, that you can do that please, still not a git pro...

AureRita commented:
Hi @Shoprunners 

Thank you for your PR, I tested your PR and, from what I understand, I shouldn't have "row" on the header if I take your module, but currently, on the inspector, I see that doesn't change anything, as you can see : 

what the inspector show with your pr and your module : 
![image](https://user-images.githubusercontent.com/100699445/168845401-01367511-00b0-4d0b-b3df-bf8630e84f2e.png)

what I have with the branch develop : 
![image](https://user-images.githubusercontent.com/100699445/168845550-2dbf9057-6cb7-4877-8cc2-d6360dfcd50b.png)

As you can see, I always have 12 occurences of the word "row". 

If this is not the good way to test your PR, could you explain to me how can I test it correctly ? 

Thank you

Shoprunners commented:
> Thank you for your PR, I tested your PR and, from what I understand, I shouldn't have "row" on the header if I take your module, but currently, on the inspector, I see that doesn't change anything, as you can see :
> Thank you

? with this PR change it is now possible to not export the header row (column names) in a csv file. It has nothing to do with html code.

AureRita commented:
I focused on the word " header row" and think it's on the inspector, 
I tested it on the export of the customers to see if the header row is register on the csv files, and as you can see, the header top is currently register : 

https://user-images.githubusercontent.com/100699445/168851759-a9b331a0-b4cc-4cbb-bfa2-d04df4fb0966.mp4

The download of the csv is launched without ask me to skip the header row.
I see on the details module that speak about adresses, but we currently can export the adresses of prestashop.
If this time I take the bad way, can you refer me the good way ?

Shoprunners commented:
Hi,
the test - module adds a new export CSV DropDown item to the orders table, not the customers  grid.
When you execute that item, a csv download will start without adding the header row to the csv file.

AureRita commented:
HI @Shoprunners 

Sorry for the waiting time, so after your answer I tested to exported the orders and add it after to see if the first row is keept, as you can see, that doesn't seems to works : 

https://user-images.githubusercontent.com/100699445/169316233-10074833-1b9e-4ef6-b39a-bd89a766b61c.mp4

Shoprunners commented:
Hi,
I think there is still a misunderstanding. Please go to the orders grid, as in the video, and do not choose export orders but the newly ( by my test module)  created entry "Export Address CSV". Maybe you have to adjust the test module a bit since var names changed during validation process.

This special export does not have a header row. The whole thing about this PR is to add the possibility for pogrammers to skip producing the header row within the PrestaSHop built in CSVResponse class.


Seriously. This was just a very small enhancement of the existing functionality and it already takes weeks to approve this. Not your fault, but from contributor site, i don't know if I want to add the next PR "to let programmers adjust the delimiters of a CSV Response".

AureRita commented:
Sorry for this misunderstanding, Indeed, as you say, we need to change a little bit the module because when I want to "Export Address CSV" I got this exception : 
![image](https://user-images.githubusercontent.com/100699445/169326853-7274b046-6358-48c6-929b-69eb1462419b.png)

Can you help me to have another module that can works or tell me the line to modify ? 

Don't let the delay stop your creativity, we are so proud of our contributors who help us move the project forward

Shoprunners commented:
So the test module was made before several adjustments. 

Please change the code in the module

from "setExport" to "setIncludeHeaderRow" in the file pointed out in your error message.

AureRita reviewed (approved):
Thank you for the time past on this PR ! It works ! as we can see on this video : 

https://user-images.githubusercontent.com/100699445/169331400-d1c12756-3f07-470c-81f8-744e4b918769.mp4


Moreover I already done an automated regression test and it works too.

Progi1984 commented:
Thanks @Shoprunners, @atomiix & @AureRita
